core/controllers/classifier.py

```diff
@@ -0,0 +1,118 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for communicating with the VM for training classifiers."""
+
+import hashlib
+import hmac
+import json
+
+from core.controllers import base
+from core.domain import classifier_domain
+from core.domain import classifier_services
+from core.domain import config_domain
+
+import feconf
+
+
+def verify_signature(message, vm_id, received_signature):
+    """Function that checks if the signature received from the VM is valid.
+
+    Args:
+        message: str. The string encoding of the dict that contains job_id and
+            classifier_data.
+        vm_id: str. The ID of the VM instance.
+        received_signature: str. The signature received from the VM.
+
+    Returns:
+        bool.
+
+    Raises:
+        UnauthorizedUserException.
+    """
+    secret = None
+    for val in config_domain.VMID_SHARED_SECRET_KEY_MAPPING.value:
+        if val['vm_id'] == vm_id:
+            secret = val['shared_secret_key'][0]
+            break
+    if secret is None:
+        return False
+    secret = str([val['shared_secret_key'] for val in (
+        config_domain.VMID_SHARED_SECRET_KEY_MAPPING.value) if val[
+            'vm_id'] == vm_id][0])
+    generated_signature = hmac.new(
+        secret, message, digestmod=hashlib.sha256).hexdigest()
+    if generated_signature != received_signature:
+        return False
+    if vm_id == feconf.DEFAULT_VM_ID and not feconf.DEV_MODE:
+        return False
+    return True
+
+
+class TrainedClassifierHandler(base.BaseHandler):
+    """This handler stores the result of the training job in datastore and
+    updates the status of the job.
+    """
+
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
+
+    def post(self):
+        """Handles POST requests."""
+        payload = self.request.get('payload')
+        payload = json.loads(payload)
+        signature = payload.get('signature')
+        message = payload.get('message')
+        vm_id = payload.get('vm_id')
+        message = json.dumps(message, sort_keys=True)
+        if verify_signature(message, vm_id, signature):
+            job_id = self.payload.get('message').get('job_id')
+            classifier_data = self.payload.get('message').get('classifier_data')
+            classifier_training_job = (
+                classifier_services.get_classifier_training_job_by_id(job_id))
+            state_name = classifier_training_job.state_name
+            exp_id = classifier_training_job.exp_id
+            exp_version = classifier_training_job.exp_version
+            algorithm_id = classifier_training_job.algorithm_id
+
+            data_schema_version = None
+            for algorithm_details in (
+                    feconf.INTERACTION_CLASSIFIER_MAPPING.values()):
+                if algorithm_details['algorithm_id'] == algorithm_id:
```

Comment:
Right -- what I'm asking is, could it be part of the job?

It's conceivable that, in the future, two different interactions use the same algorithm. It seems to me that it would make more sense to just look up the correct interaction id and algorithm id directly, rather than searching through the list of all possible algorithms. Especially since we already know the interaction ID when the job is created.